lib/jsdom/living/traversal/TreeWalker-impl.js

```diff
@@ -0,0 +1,197 @@
+"use strict";
+
+const DOMException = require("../../web-idl/DOMException");
+const idlUtils = require("../generated/utils");
+
+//FIXME: Once NodeFilter is ported to IDL method, uncomment these.
+const FILTER_ACCEPT = 1; //NodeFilter.FILTER_ACCEPT;
+const FILTER_REJECT = 2; //NodeFilter.FILTER_REJECT;
+const FILTER_SKIP = 3; //NodeFilter.FILTER_SKIP;
+const FIRST = false;
+const LAST = true;
+const NEXT = false;
+const PREVIOUS = true;
+
+function isNull(o) {
+	return o === null || typeof o === "undefined";
+}
+
+class TreeWalkerImpl {
+	_filterNode(node) {
+		const n = node.nodeType - 1;
+		if (!((1 << n) & this.whatToShow)) return FILTER_SKIP;
+		const filter = this.filter;
+		if (isNull(filter)) return FILTER_ACCEPT;
+		let result;
+
+		if (typeof filter === "function") result = filter(idlUtils.wrapperForImpl(node));  //non-standard
+		else result = filter.acceptNode(idlUtils.wrapperForImpl(node));
+
+		if (result === true) result = FILTER_ACCEPT;  //non-standard
+		return result;
+	};
+
+	_traverseChildren(type) {
+		let node = this._currentNode;
+		node = (type === FIRST ? node.firstChild : node.lastChild);
+		if (isNull(node)) return null;
+
+		main: while (true) {
+			const result = this._filterNode(node);
+			if (result === FILTER_ACCEPT) return (this._currentNode = node);
+
+			if (result === FILTER_SKIP) {
+				const child = (type === FIRST ? node.firstChild : node.lastChild);
+
+				if (!isNull(child)) {
+					node = child;
+					continue main;
+				}
+			}
+
+			while (true) {
+				const sibling = (type === FIRST ? node.nextSibling : node.previousSibling);
```

Comment:
Yeah, it's against this project's style though. I'll try to make lint enforce it.